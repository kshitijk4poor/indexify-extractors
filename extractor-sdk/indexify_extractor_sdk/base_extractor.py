import json
import os
from importlib import import_module
import logging
from indexify.extractor_sdk import Content, Feature, Extractor, ExtractorMetadata, EmbeddingSchema
from typing import (
    Dict,
    List,
    Tuple,
    Type,
    Union,
    get_type_hints,
)

import requests
from genson import SchemaBuilder
from pydantic import BaseModel, Field, Json

EXTRACTORS_PATH = os.path.join(os.path.expanduser("~"), ".indexify-extractors")
EXTRACTORS_MODULE = "indexify_extractors"
EXTRACTOR_MODULE_PATH = os.path.join(EXTRACTORS_PATH, EXTRACTORS_MODULE)


def load_extractor(name: str) -> Tuple[Extractor, Type[BaseModel]]:
    module_name, class_name = name.split(":")
    wrapper = ExtractorWrapper(module_name, class_name)
    return (wrapper._instance, wrapper._param_cls)


class ExtractorWrapper:
    def __init__(self, module_name: str, class_name: str):
        module = import_module(module_name)
        cls = getattr(module, class_name)
        self._instance: Extractor = cls()
        self._param_cls = get_type_hints(cls.extract).get("params", None)
        extract_batch = getattr(self._instance, "extract_batch", None)
        self._has_batch_extract = True if callable(extract_batch) else False

    def _param_from_json(self, param: Json) -> BaseModel:
        if self._param_cls is None:
            return {}

        param_dict = {}
        if param is not None and param != "null":
            param_dict = json.loads(param)

        try:
            return self._param_cls.model_validate(param_dict)
        except Exception as e:
            print(f"Error validating input params: {e}")

        return {}

    def extract_batch(
        self, content_list: Dict[str, Content], input_params: Dict[str, Json]
    ) -> Dict[str, List[Union[Feature, Content]]]:
        if self._has_batch_extract:
            task_ids = []
            task_contents = []
            params = []
            for task_id, content in content_list.items():
                param_instance = self._param_from_json(input_params.get(task_id, None))
                params.append(param_instance)
                task_ids.append(task_id)
                task_contents.append(content)

            try:
                result = self._instance.extract_batch(task_contents, params)
            except Exception as e:
                logging.error(f"Error extracting content: {e}")
                raise e
            out: Dict[str, List[Union[Feature, Content]]] = {}
            for i, extractor_out in enumerate(result):
                out[task_ids[i]] = extractor_out
            return out
        out = {}
        for task_id, content in content_list.items():
            param_instance = self._param_from_json(input_params.get(task_id, None))
            out[task_id] = self._instance.extract(content, param_instance)
        return out

    def describe(self) -> ExtractorMetadata:
        s_input = self._instance.sample_input()
        input_params = None
        if type(s_input) == tuple:
            (s_input, input_params) = s_input
        # Come back to this when we can support schemas based on user defined input params
        if input_params is None:
            input_params = (
                self._param_cls().model_dump_json()
                if self._param_cls is not None
                else None
            )
        outputs: Dict[str, List[Union[Feature, Content]]] = self.extract_batch(
            {"task_id": s_input},
            {"task_id": input_params},
        )
        embedding_schemas = {}
        json_schema = (
            self._param_cls.model_json_schema() if self._param_cls is not None else None
        )
        output = outputs["task_id"]
        for out in output:
            features = out.features if type(out) == Content else [out]
            for feature in features:
                if feature.feature_type == "embedding":
                    embedding_schema = EmbeddingSchema(
                        dim=len(feature.value["values"]),
                        distance=feature.value["distance"],
                    )
                    embedding_schemas[feature.name] = embedding_schema
        return ExtractorMetadata(
            name=self._instance.name,
            version=self._instance.version,
            description=self._instance.description,
            python_dependencies=self._instance.python_dependencies,
            system_dependencies=self._instance.system_dependencies,
            embedding_schemas=embedding_schemas,
            metadata_schemas={},
            input_mime_types=self._instance.input_mime_types,
            input_params=json_schema,
        )
