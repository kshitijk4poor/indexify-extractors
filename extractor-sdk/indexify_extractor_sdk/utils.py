from itertools import islice
import requests
import platform
import fsspec
import json
from rich.console import Console

console = Console()

# https://docs.python.org/3/library/itertools.html#itertools.batched
def batched(iterable, n):
    # batched('ABCDEFG', 3) --> ABC DEF G
    if n < 1:
        raise ValueError("n must be at least one")
    it = iter(iterable)
    while batch := tuple(islice(it, n)):
        yield batch


def log_event(event, value):
    try:
        requests.post(
            "https://getindexify.ai/api/analytics", json={"event": event, "value": value, "platform":platform.platform(), "machine": platform.machine()}
        , timeout=1)
    except Exception as e:
        # fail silently
        pass

def read_extractors_json_file():
    file_path = f's3://indexifyextractors/indexify-extractors/extractors.json'

    fs = fsspec.filesystem('s3', anon=True)

    with fs.open(file_path, "r") as file:
        # Load the JSON content from the file
        json_content = json.load(file)

    return json_content

def extractors_by_name():
    extractors_info_list = read_extractors_json_file()
    result = {}
    for extractor_info in extractors_info_list:
        result[extractor_info["name"]] = extractor_info
    return result


class ExtractorIndex:

    def __init__(self) -> None:
        file_path = f's3://indexifyextractors/indexify-extractors/extractors.json'
        fs = fsspec.filesystem('s3', anon=True)
        self._index = {}
        try:
            with fs.open(file_path, "r") as file:
                # Load the JSON content from the file
                json_index = json.load(file)
                for extractor_info in json_index:
                    self._index[extractor_info["name"]] = extractor_info
        except Exception as e:
            value = {"stage": "extractor_list", "error": str(e)}
            log_event("extractor_download_failed", value=value)
            raise e

    def metadata_by_name(self, name):
        try:
            return self._index.get(name)
        except Exception as e:
            value = {"stage": "extractor_metadata", "error": str(e)}
            log_event("extractor_download_failed", value=value)
            console.print(f"[bold #f04318]Extractor {name} not found[/]")
            console.print(f"[bold #f04318]Use command: [yellow]indexify-extractor list[/yellow] to see the list of available extractors[/]")
            raise e
