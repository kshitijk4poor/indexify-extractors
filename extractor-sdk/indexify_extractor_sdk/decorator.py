import inspect
from typing import Callable, List, Optional, get_type_hints

from .base_extractor import Content, Extractor


def extractor(
    name: Optional[str] = None,
    description: Optional[str] = "",
    version: Optional[str] = "",
    python_dependencies: Optional[List[str]] = None,
    system_dependencies: Optional[List[str]] = None,
    input_mime_types: Optional[List[str]] = None,
    sample_content: Optional[Callable] = None,
):
    args = locals()
    del args["sample_content"]

    def construct(fn):
        hint = get_type_hints(fn).get("params", dict)

        if not args.get("name"):
            args[
                "name"
            ] = f"{inspect.getmodule(inspect.stack()[1][0]).__name__}:{fn.__name__}"

        class DecoratedFn(Extractor):
            def extract(self, content: Content, params: hint) -> List[Content]:  # type: ignore
                return fn(content, params)

            def sample_input(self) -> Content:
                return sample_content() if sample_content else self.sample_text()

        for key, val in args.items():
            setattr(DecoratedFn, key, val)

        return DecoratedFn

    return construct
