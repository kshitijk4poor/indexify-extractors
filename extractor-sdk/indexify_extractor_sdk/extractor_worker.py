import concurrent.futures
from typing import List, Dict, Any
from .base_extractor import ExtractorWrapper
from pydantic import Json
import concurrent
from .downloader import ExtractorMetadataStore
from indexify.extractor_sdk import ExtractorMetadata
import asyncio
from concurrent.futures.process import BrokenProcessPool


# str here is ExtractorDescription.name
extractor_wrapper_map: Dict[str, ExtractorWrapper] = {}

# List of ExtractorDescription
# This is used to report the available extractors to the coordinator
extractor_descriptions: Dict[str, ExtractorMetadata] = {}

def load_extractors(name: str, extractor_module_class: str):
    """Load an extractor to the memory: extractor_wrapper_map."""
    global extractor_wrapper_map
    if name in extractor_wrapper_map:
        return
    extractor_id = f"indexify_extractors.{extractor_module_class}"
    extractor_wrapper = ExtractorWrapper.from_name(extractor_id)
    extractor_wrapper_map[name] = extractor_wrapper


class ExtractorWorker:
    def __init__(self, extractor_metadata_store: ExtractorMetadataStore, workers: int=1) -> None:
        self._extractor_metadata_store = extractor_metadata_store
        self._extractors = {}
        self._executor: concurrent.futures.ProcessPoolExecutor = concurrent.futures.ProcessPoolExecutor(max_workers=workers)

    def submit(self, extractor: str, inputs: Dict[str, Any], params: Dict[str, Any]) -> Dict[str, List[Any]]:
        extractor_module_class = self._extractor_metadata_store.extractor_module_class(extractor)
        return self._executor.submit(_extract_content, extractor, extractor_module_class, inputs, params, extractor)
    
    async def async_submit(self, extractor: str, inputs: Dict[str, Any], params: Dict[str, Any]) -> Dict[str, List[Any]]:
        extractor_module_class = self._extractor_metadata_store.extractor_module_class(extractor)
        try:
            resp = await asyncio.get_running_loop().run_in_executor(
            self._executor,
            _extract_content,
            extractor,
            extractor_module_class,
            inputs,
            params,
            extractor,
            )
        except BrokenProcessPool as mp:
            self._executor.shutdown(wait=True, cancel_futures=True)
            raise mp
        return resp

def _extract_content(
        extractor: str,
        extractor_module_class: str,
        inputs: Dict[str, Any],
        params: Dict[str, Json],
    ) -> Dict[str, List[Any]]:
    if extractor not in extractor_wrapper_map:
        load_extractors(extractor, extractor_module_class)
        
    extractor_wrapper: ExtractorWrapper = extractor_wrapper_map[extractor]
    task_ids, extractor_inputs, extractor_params = [], [], []
    for task_id, content in inputs.items():
        task_ids.append(task_id)
        extractor_inputs.append(content)
        extractor_params.append(params[task_id])

    results = extractor_wrapper.extract_batch(inputs, params)
    output = {}
    for (task_id, output) in zip(task_ids, results):
        output[task_id] = output
    return output
