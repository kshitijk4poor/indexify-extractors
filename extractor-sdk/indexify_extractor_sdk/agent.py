import ssl
import yaml
import asyncio
from . import coordinator_service_pb2
from .coordinator_service_pb2_grpc import CoordinatorServiceStub
import grpc
import json
from typing import List, Dict, Union, Optional
from .base_extractor import ExtractorDescription
from .base_extractor import Content, Feature, Embedding
from .content_downloader import (
    create_content,
    download_content,
    UrlConfig,
)
from .extractor_worker import (
    extract_content,
    create_executor,
    describe,
)
from concurrent.futures.process import BrokenProcessPool
from .ingestion_api_models import (
    ApiContent,
    ApiFeature,
    BeginExtractedContentIngest,
    ExtractedFeatures,
    FinishExtractedContentIngest,
    ApiBeginExtractedContentIngest,
    ApiExtractedFeatures,
    ApiFinishExtractedContentIngest,
    ApiBeginMultipartContent,
    BeginMultipartContent,
    ApiFinishMultipartContent,
    ApiMultipartContentFrame,
    MultipartContentFrame,
    FinishMultipartContent,
)
from .utils import batched
from .server import http_server, ServerRouter, get_server_advertise_addr
import concurrent
import websockets
from .task_store import TaskStore, CompletedTask
from websockets.exceptions import ConnectionClosed
from pydantic import BaseModel, Json

CONTENT_FRAME_SIZE = 1024 * 1024

MAX_MESSAGE_LENGTH = 16 * 1024 * 1024

# maximum number of content in extractor call
DEFAULT_BATCH_SIZE = 10


def begin_message(task_outcome, task: coordinator_service_pb2.Task, _executor_id):
    return ApiBeginExtractedContentIngest(
        BeginExtractedContentIngest=BeginExtractedContentIngest(
            task_id=task_outcome.task_id,
            executor_id=_executor_id,
            task_outcome=task_outcome.task_outcome,
        )
    )


async def send_extracted_content(ws, content: ApiContent, id: int, frame_size):
    # start new multipart content
    await ws.send(
        ApiBeginMultipartContent(
            BeginMultipartContent=BeginMultipartContent(id=id)
        ).model_dump_json()
    )

    # send data in chunks of frame_size
    for i in range(0, len(content.bytes), frame_size):
        slice = content.bytes[i : i + frame_size]
        content_frame = ApiMultipartContentFrame(
            MultipartContentFrame=MultipartContentFrame(bytes=slice)
        )
        await ws.send(content_frame.model_dump_json())

    # finish multipart content with features
    await ws.send(
        ApiFinishMultipartContent(
            FinishMultipartContent=FinishMultipartContent(
                content_type=content.content_type,
                features=content.features,
                labels=content.labels,
            )
        ).model_dump_json()
    )


class TaskReportError(Exception):
    """Exception raised for errors in the task reporting process."""

    def __init__(self, task_id, message="Failed to report task"):
        self.task_id = task_id
        self.message = message
        super().__init__(self.message)

    def __str__(self):
        return f"{self.message}"


async def process_task_outcome(
    task_outcome: CompletedTask,
    task: coordinator_service_pb2.Task,
    url,
    _executor_id,
    ssl_context,
    frame_size=CONTENT_FRAME_SIZE,
):
    try:
        async with websockets.connect(
            url, ssl=ssl_context, ping_interval=5, ping_timeout=30
        ) as ws:
            # start new extracted content ingest
            await ws.send(
                begin_message(task_outcome, task, _executor_id).model_dump_json()
            )

            num_extracted_content = 0
            if task_outcome.task_outcome == "Success":
                num_extracted_content = len(task_outcome.new_content)
                # send all contents one at a time
                for i, content in enumerate(task_outcome.new_content):
                    await send_extracted_content(
                        ws, content, id=i + 1, frame_size=frame_size
                    )

                # send all features one at a time
                for feature in task_outcome.features:
                    extracted_features = ApiExtractedFeatures(
                        ExtractedFeatures=ExtractedFeatures(
                            content_id=task.content_metadata.id, features=[feature]
                        )
                    )
                    await ws.send(extracted_features.model_dump_json())
                print(
                    f"finished message {FinishExtractedContentIngest(num_extracted_content=len(task_outcome.new_content)).model_dump_json()}"
                )

            # finish extracted content ingest
            finish_msg = ApiFinishExtractedContentIngest(
                FinishExtractedContentIngest=FinishExtractedContentIngest(
                    num_extracted_content=num_extracted_content
                )
            )

            await ws.send(finish_msg.model_dump_json())

            response = await ws.recv()
            response_data = json.loads(response)
            print(f"response: {response_data}")
            if "Error" in response_data:
                raise TaskReportError(task_outcome.task_id, response_data["Error"])

    except ConnectionClosed as e:
        if not e.rcvd is None:
            # the connection was closed by the server with an error message
            raise TaskReportError(
                task_outcome.task_id,
                f"Connection closed with code {e.code} reason {e.reason}",
            )
        else:
            # otherwise abnormal close, retry
            raise e


class ContentBatch(BaseModel):
    extractor: str
    content_list: Dict[str, Content] = {}
    params: Dict[str, Json] = {}
    extractors: Dict[str, str] = {}


class ExtractorState(BaseModel):
    pending_batches: int = 0
    new_batches: List[ContentBatch]


def extractor_state_new(extractor: str) -> ExtractorState:
    return ExtractorState(new_batches=[ContentBatch(extractor=extractor)])


class ExtractTask(asyncio.Task):
    def __init__(self, *, executor, content_batch: ContentBatch, **kwargs):
        kwargs["name"] = "extract_content"
        kwargs["loop"] = asyncio.get_event_loop()
        super().__init__(
            extract_content(
                loop=asyncio.get_running_loop(),
                executor=executor,
                content_list=content_batch.content_list,
                params=content_batch.params,
                extractors=content_batch.extractors,
            ),
            **kwargs,
        )
        self.content_batch = content_batch


class ExtractorAgent:
    def __init__(
        self,
        executor_id: str,
        extractors: List[coordinator_service_pb2.Extractor],
        coordinator_addr: str,
        executor: concurrent.futures.ProcessPoolExecutor,
        num_workers,
        extractor_arg,
        listen_port: int,
        advertise_addr: Optional[str],
        ingestion_addr: str = "localhost:8900",
        config_path: Optional[str] = None,
        download_method: str = "direct",
        batch_size: int = DEFAULT_BATCH_SIZE,
    ):
        self.num_workers = num_workers
        self.extractor_arg = extractor_arg
        self._use_tls = False
        if config_path:
            with open(config_path, "r") as f:
                config = yaml.safe_load(f)
                self._config = config
            if config.get("use_tls", False):
                print("Running the extractor with TLS enabled")
                self._use_tls = True
                tls_config = config["tls_config"]
                self._ssl_context = ssl.create_default_context(
                    ssl.Purpose.SERVER_AUTH, cafile=tls_config["ca_bundle_path"]
                )
                self._ssl_context.load_cert_chain(
                    certfile=tls_config["cert_path"], keyfile=tls_config["key_path"]
                )
                self._protocol = "wss"
                self._tls_config = tls_config
            else:
                self._ssl_context = None
                self._protocol = "ws"
        else:
            self._ssl_context = None
            self._protocol = "ws"
            self._config = {}

        self._task_store: TaskStore = TaskStore()
        self._executor_id = executor_id
        self._extractors = extractors
        self._has_registered = False
        self._coordinator_addr = coordinator_addr
        self._ingestion_addr = ingestion_addr
        self._listen_port = listen_port
        self._advertise_addr = advertise_addr
        self._executor = executor
        self._download_method = download_method
        self._batch_size = batch_size

    async def ticker(self):
        while True:
            await asyncio.sleep(5)
            yield coordinator_service_pb2.HeartbeatRequest(
                executor_id=self._executor_id,
                pending_tasks=self._task_store.num_pending_tasks(),
            )

    async def register(self):
        # This needs to be here because every time we register we need to create a new channel
        # because the old one might have been closed after hb was broken or we could never connect
        if self._use_tls:
            # Load the certificate and key files
            with open(self._config["tls_config"]["cert_path"], "rb") as f:
                cert = f.read()
            with open(self._config["tls_config"]["key_path"], "rb") as f:
                key = f.read()
            with open(self._config["tls_config"]["ca_bundle_path"], "rb") as f:
                ca_cert = f.read()
            credentials = grpc.ssl_channel_credentials(
                root_certificates=ca_cert, private_key=key, certificate_chain=cert
            )
            self._channel = grpc.aio.secure_channel(
                self._coordinator_addr,
                credentials=credentials,
                options=[
                    ("grpc.max_send_message_length", MAX_MESSAGE_LENGTH),
                    ("grpc.max_receive_message_length", MAX_MESSAGE_LENGTH),
                ],
            )
        else:
            self._channel = grpc.aio.insecure_channel(
                self._coordinator_addr,
                options=[
                    ("grpc.max_send_message_length", MAX_MESSAGE_LENGTH),
                    ("grpc.max_receive_message_length", MAX_MESSAGE_LENGTH),
                ],
            )

        self._stub: CoordinatorServiceStub = CoordinatorServiceStub(self._channel)
        req = coordinator_service_pb2.RegisterExecutorRequest(
            executor_id=self._executor_id,
            addr=self._advertise_addr,
            extractors=self._extractors,
        )
        return await self._stub.RegisterExecutor(req)

    async def task_completion_reporter(self):
        print("starting task completion reporter")
        # We should copy only the keys and not the values
        url = f"{self._protocol}://{self._ingestion_addr}/write_content"
        while True:
            outcomes = await self._task_store.task_outcomes()
            for task_outcome in outcomes:
                print(
                    f"reporting outcome of task {task_outcome.task_id}, outcome: {task_outcome.task_outcome}, num_content: {len(task_outcome.new_content)}, num_features: {len(task_outcome.features)}"
                )
                task: coordinator_service_pb2.Task = self._task_store.get_task(
                    task_outcome.task_id
                )
                try:
                    await process_task_outcome(
                        task_outcome, task, url, self._executor_id, self._ssl_context
                    )
                except TaskReportError as e:
                    print(f"failed to report task {e.task_id}, exception: {e}")
                    self._task_store.report_failed(task_id=e.task_id)
                    continue
                except Exception as e:
                    # the connection was dropped in the middle of the reporting process, retry
                    print(
                        f"failed to report task {task_outcome.task_id}, exception: {e}, retrying"
                    )
                    continue

                self._task_store.mark_reported(task_id=task_outcome.task_id)

    def get_url_config(self, task: coordinator_service_pb2.Task) -> UrlConfig:
        if self._download_method == "server-proxy":
            protocol = "https://" if self._config.get("use_tls") else "http://"
            url = f"{protocol}{self._ingestion_addr}/namespaces/{task.content_metadata.namespace}/content/{task.content_metadata.id}/download"
            return UrlConfig(url=url, config=self._config)
        else:
            return UrlConfig(url=task.content_metadata.storage_url, config={})

    async def task_launcher(self):
        async_tasks = []
        extractor_states = {}
        async_tasks.append(
            asyncio.create_task(
                self._task_store.get_runnable_tasks(), name="get_runnable_tasks"
            )
        )
        while True:
            for extractor, state in extractor_states.items():
                if (
                    state.pending_batches == 0
                    and len(state.new_batches[0].content_list) != 0
                ):
                    content_batch = state.new_batches.pop(0)
                    print(
                        f"extracting content for {extractor} tasks {len(content_batch.content_list.keys())} {content_batch.content_list.keys()}"
                    )
                    async_tasks.append(
                        ExtractTask(
                            executor=self._executor,
                            content_batch=content_batch,
                        )
                    )
                    if len(state.new_batches) == 0:
                        state.new_batches.append(ContentBatch(extractor=extractor))
                    state.pending_batches += 1

            done, pending = await asyncio.wait(
                async_tasks, return_when=asyncio.FIRST_COMPLETED
            )
            async_tasks = list(pending)
            for async_task in done:
                if async_task.get_name() == "get_runnable_tasks":
                    result = await async_task
                    for _, task in result.items():
                        url_config = self.get_url_config(task)
                        async_tasks.append(
                            asyncio.create_task(
                                download_content(task.id, url_config),
                                name="download_content",
                            )
                        )
                    async_tasks.append(
                        asyncio.create_task(
                            self._task_store.get_runnable_tasks(),
                            name="get_runnable_tasks",
                        )
                    )
                elif async_task.get_name() == "download_content":
                    # Process all completed downloads and accumulate them in batches
                    # without creating extraction tasks right away.
                    task_id, bytes = await async_task
                    if isinstance(bytes, Exception):
                        print(f"failed to download content {bytes} for task {task_id}")
                        completed_task = CompletedTask(
                            task_id=task_id,
                            task_outcome="Failed",
                            new_content=[],
                            features=[],
                        )
                        self._task_store.complete(outcome=completed_task)
                        continue
                    task = self._task_store.get_task(task_id)
                    state = extractor_states.setdefault(
                        task.extractor, extractor_state_new(task.extractor)
                    )
                    if len(state.new_batches[-1].content_list) == self._batch_size:
                        state.new_batches.append(ContentBatch(extractor=task.extractor))
                    content_batch = state.new_batches[-1]
                    content_batch.content_list[task_id] = create_content(bytes, task)
                    content_batch.params[task_id] = task.input_params
                    content_batch.extractors[task_id] = task.extractor
                elif async_task.get_name() == "extract_content":
                    content_batch = async_task.content_batch
                    state = extractor_states[content_batch.extractor]
                    state.pending_batches -= 1
                    try:
                        outputs = await async_task
                        for task_id, e_output in outputs.items():
                            print(f"completed task {task_id}")
                            new_content: List[ApiContent] = []
                            new_features: List[ApiFeature] = []
                            out: Union[Feature, Content]
                            for out in e_output:
                                if type(out) == Feature:
                                    new_features.append(
                                        ApiFeature.from_feature(feature=out)
                                    )
                                    continue
                                new_content.append(ApiContent.from_content(content=out))
                            completed_task = CompletedTask(
                                task_id=task_id,
                                task_outcome="Success",
                                new_content=new_content,
                                features=new_features,
                            )
                            self._task_store.complete(outcome=completed_task)
                    except BrokenProcessPool as bp:
                        print(f"failed to execute tasks {bp}, retrying")
                        self._executor.shutdown(wait=True, cancel_futures=True)
                        self._executor = create_executor(
                            workers=self.num_workers, extractor_id=self.extractor_arg
                        )
                        for task_id in async_task.content_batch.content_list.keys():
                            self._task_store.retriable_failure(task_id)
                        continue
                    except Exception as e:
                        task_ids = ",".join(
                            async_task.content_batch.content_list.keys()
                        )
                        print(f"failed to execute tasks {task_ids} {e}")
                        for task_id in async_task.content_batch.content_list.keys():
                            completed_task = CompletedTask(
                                task_id=task_id,
                                task_outcome="Failed",
                                new_content=[],
                                features=[],
                            )
                            self._task_store.complete(outcome=completed_task)
                        continue

    async def run(self):
        import signal

        asyncio.get_event_loop().add_signal_handler(
            signal.SIGINT, self.shutdown, asyncio.get_event_loop()
        )
        server_router = ServerRouter(self._executor)
        self._http_server = http_server(server_router, port=self._listen_port)
        asyncio.create_task(self._http_server.serve())
        if not self._advertise_addr:
            self._advertise_addr = await get_server_advertise_addr(self._http_server)
        print(f"advertise addr is {self._advertise_addr}")
        asyncio.create_task(self.task_launcher())
        asyncio.create_task(self.task_completion_reporter())
        self._should_run = True
        while self._should_run:
            print("attempting to register")
            try:
                await self.register()
                self._has_registered = True
            except Exception as e:
                print(f"failed to register: {e}")
                await asyncio.sleep(5)
                continue
            hb_ticker = self.ticker()
            print("starting heartbeat")
            try:
                hb_response_it = self._stub.Heartbeat(hb_ticker)
                resp: coordinator_service_pb2.HeartbeatResponse
                async for resp in hb_response_it:
                    self._task_store.add_tasks(resp.tasks)
            except Exception as e:
                print(f"failed to heartbeat{e}")
                continue

    async def _shutdown(self, loop):
        print("shutting down agent ...")
        self._should_run = False
        self._http_server.should_exit = True
        await self._channel.close()
        for task in asyncio.all_tasks(loop):
            task.cancel()

    def shutdown(self, loop):
        self._executor.shutdown(wait=True, cancel_futures=True)
        loop.create_task(self._shutdown(loop))
