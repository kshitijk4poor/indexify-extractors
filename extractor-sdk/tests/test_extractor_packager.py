import gzip
import io
import tarfile

import pytest
from indexify_extractor_sdk.packager import ExtractorPackager


@pytest.fixture
def packager():
    return ExtractorPackager(
        module_name="indexify_extractor_sdk.mock_extractor",
        class_name="MockExtractor",
        dockerfile_template_path="dockerfiles/Dockerfile.extractor",
        verbose=False,
        dev=False,
        gpu=False,
    )


@pytest.fixture
def packager_dev():
    return ExtractorPackager(
        module_name="indexify_extractor_sdk.mock_extractor",
        class_name="MockExtractor",
        dockerfile_template_path="dockerfiles/Dockerfile.extractor",
        verbose=False,
        dev=True,
        gpu=False,
    )


def test_generate_dockerfile(packager, snapshot):
    # `syrupy` snapshot testing -
    dockerfile_content = packager._generate_dockerfile()
    assert dockerfile_content == snapshot


def test_generate_dockerfile_dev(packager_dev, snapshot):
    dockerfile_content = packager_dev._generate_dockerfile()
    assert dockerfile_content == snapshot


def test_generate_compressed_tarball(packager):
    dockerfile_content = packager._generate_dockerfile()
    tarball_bytes = packager._generate_compressed_tarball(dockerfile_content)

    with gzip.open(io.BytesIO(tarball_bytes), "rb") as f:
        with tarfile.open(fileobj=f, mode="r") as tar:
            tarball = tar.getnames()

    # print contents of tarball
    assert "Dockerfile" in tarball
    assert "mock_extractor.py" in tarball

    # dev files are NOT there
    assert "pyproject.toml" not in tarball
    assert "README.md" not in tarball
    assert "indexify_extractor_sdk/__init__.py" not in tarball
    assert "indexify_extractor_sdk/base_extractor.py" not in tarball


def test_generate_compressed_tarball_dev(packager_dev):
    dockerfile_content = packager_dev._generate_dockerfile()
    tarball_bytes = packager_dev._generate_compressed_tarball(dockerfile_content)

    with gzip.open(io.BytesIO(tarball_bytes), "rb") as f:
        with tarfile.open(fileobj=f, mode="r") as tar:
            tarball = tar.getnames()

    # print contents of tarball
    assert "Dockerfile" in tarball
    assert "mock_extractor.py" in tarball

    # minimum dev files are there
    assert "pyproject.toml" in tarball
    assert "README.md" in tarball
    assert "indexify_extractor_sdk/__init__.py" in tarball
    assert "indexify_extractor_sdk/base_extractor.py" in tarball


@pytest.mark.skip(reason="This test will build the docker image and will take time")
def test_package(packager):
    # This test will create an image at mock_extractor:latest
    # Enter the image with: `docker run -it --entrypoint "" mock_extractor:latest sh -c /bin/bash`
    packager.package()
