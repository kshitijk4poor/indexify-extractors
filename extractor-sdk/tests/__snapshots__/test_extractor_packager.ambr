# serializer version: 1
# name: test_generate_dockerfile
  '''
  FROM --platform=linux/amd64 ubuntu:22.04

  RUN apt-get update -y && \
      apt-get -y install lsb-release ca-certificates

  RUN update-ca-certificates

  RUN echo "deb [trusted=yes] https://cf-repo.diptanu-6d5.workers.dev/repo $(lsb_release -cs) main" | tee /etc/apt/sources.list.d/indexify-repo.list && \
      apt-get update -y && \
      apt-get install -y lsb-release libssl-dev apt-transport-https && \
      apt-get install -y python3.11 python3.11-dev python3-pip python3.11-venv && \
      apt-get install -y  sl cowsay && \
      apt-get -y clean

  # SET ENV VAR EXTRACTOR_PATH. This is used by the indexify-extractor join command to find the extractor
  ENV EXTRACTOR_PATH mock_extractor:MockExtractor

  WORKDIR /indexify

  RUN python3.11 -m venv ve
  # Use the virtualenv, to make sure we're using the right Python version
  ENV PATH="/indexify/ve/bin:$PATH"

  RUN pip3 install --no-input --upgrade pip && \
      pip3 install --no-input wheel && \
      pip3 install --no-input --extra-index-url https://download.pytorch.org/whl/cpu tinytext pyfiglet


  RUN pip3 install --no-input indexify_extractor_sdk --extra-index-url https://download.pytorch.org/whl/cpu


  # Copy over the extractor as indexify_extractor_sdk.mock_extractor.py into /indexify/indexify_extractor_sdk.mock_extractor.py
  COPY mock_extractor.py /indexify/mock_extractor.py

  # verify installation
  RUN indexify-extractor --help

  ENTRYPOINT [ "indexify-extractor" ]
  '''
# ---
# name: test_generate_dockerfile_dev
  '''
  FROM --platform=linux/amd64 ubuntu:22.04

  RUN apt-get update -y && \
      apt-get -y install lsb-release ca-certificates

  RUN update-ca-certificates

  RUN echo "deb [trusted=yes] https://cf-repo.diptanu-6d5.workers.dev/repo $(lsb_release -cs) main" | tee /etc/apt/sources.list.d/indexify-repo.list && \
      apt-get update -y && \
      apt-get install -y lsb-release libssl-dev apt-transport-https && \
      apt-get install -y python3.11 python3.11-dev python3-pip python3.11-venv && \
      apt-get install -y  sl cowsay && \
      apt-get -y clean

  # SET ENV VAR EXTRACTOR_PATH. This is used by the indexify-extractor join command to find the extractor
  ENV EXTRACTOR_PATH mock_extractor:MockExtractor

  WORKDIR /indexify

  RUN python3.11 -m venv ve
  # Use the virtualenv, to make sure we're using the right Python version
  ENV PATH="/indexify/ve/bin:$PATH"

  RUN pip3 install --no-input --upgrade pip && \
      pip3 install --no-input wheel && \
      pip3 install --no-input --extra-index-url https://download.pytorch.org/whl/cpu tinytext pyfiglet


  COPY indexify_extractor_sdk /indexify/indexify_extractor_sdk

  COPY README.md /indexify/README.md
  COPY pyproject.toml /indexify/pyproject.toml

  RUN pip3 install .


  # Copy over the extractor as indexify_extractor_sdk.mock_extractor.py into /indexify/indexify_extractor_sdk.mock_extractor.py
  COPY mock_extractor.py /indexify/mock_extractor.py

  # verify installation
  RUN indexify-extractor --help

  ENTRYPOINT [ "indexify-extractor" ]
  '''
# ---
